### ISSUE TITLE
EKS Managed Node Group cannot be created, user cannot perform iam:GetRole on ressource

### ISSUE THREAD

**Issue by aozoeao** [terraform-aws-modules/terraform-aws-eks#2018]
- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]:

- Terraform version:
Terraform v0.14.11
- Provider version(s):
+ provider registry.terraform.io/hashicorp/aws v4.9.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.2.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.10.0
+ provider registry.terraform.io/hashicorp/tls v3.3.0

## Reproduction Code [Required]

- eks.tf
```
module "eks" {
  source = "git::https://github.com/terraform-aws-modules/terraform-aws-eks.git?ref=v18.20.2"

  cluster_name                         = "${var.project}-${var.env}"
  cluster_version                      = var.cluster_version
  cluster_endpoint_private_access      = true
  cluster_endpoint_public_access       = true
  cluster_endpoint_public_access_cidrs = var.my_ips
  
  vpc_id     = module.vpc.vpc_id
  subnet_ids = module.vpc.private_subnets

  manage_aws_auth_configmap   = true
  create_cloudwatch_log_group = false

  enable_irsa = true

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {
      resolve_conflicts = "OVERWRITE"
    }
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
    }
  }

  # Extend cluster security group rules
  cluster_security_group_additional_rules = {
    egress_nodes_ephemeral_ports_tcp = {
      description                = "To node 1025-65535"
      protocol                   = "tcp"
      from_port                  = 1025
      to_port                    = 65535
      type                       = "egress"
      source_node_security_group = true
    }
  }

  # Extend node-to-node security group rules
  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }
    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }
  }

  eks_managed_node_group_defaults = {
    ami_type                              = "AL2_x86_64"
    enable_monitoring                     = false
    create_launch_template                = false
    launch_template_name                  = ""
    attach_cluster_primary_security_group = true
    iam_role_attach_cni_policy            = true
  }

  eks_managed_node_groups = {
    safety = {
      name                     = "safety"
      instance_types           = ["m5.large"]
      instance_refresh_enabled = true
      min_size                 = 0
      max_size                 = 1
      labels = {
        type = "safety"
      }
      tags = {
        "k8s.io/cluster-autoscaler/enabled"                   = "true",
        "k8s.io/cluster-autoscaler/${var.project}-${var.env}" = "owned"
      }
    }
}

```
vpc.tf
```
module "vpc" {
  source = "git::git@github.com:terraform-aws-modules/terraform-aws-vpc.git?ref=v3.14.0"

  name = "${var.project}-${var.env}"
  cidr = var.cidr

  azs             = ["${var.region}a", "${var.region}b", "${var.region}c"]
  private_subnets = var.private_subnets
  public_subnets  = var.public_subnets

  enable_nat_gateway = true
  enable_vpn_gateway = true

  private_subnet_tags = {
    "kubernetes.io/role/internal-elb" = "1",
  }

  public_subnet_tags = {
    "kubernetes.io/role/elb"                          = "1",
    "kubernetes.io/cluster/${var.project}-${var.env}" = "shared"
  }

  default_vpc_enable_dns_hostnames = true
  enable_dns_hostnames             = true
}
```

I'm not using workspaces.
I have cleared the cache before trying anything.

steps to reproduce:
- terraform init
- terraform apply
After the EKS cluster was successfully built the node group was not created displaying the error below.

## Expected behavior

EKS Node Group safety is created.

## Actual behavior

EKS Node Group cannot be created, as said in the teminal output my user cannot perform iam:GetRole on the ressource. However my user is admin.

### Terminal Output Screenshot(s)

![image](https://user-images.githubusercontent.com/35883521/163399033-e2ec0dfc-6b91-4aaa-b8da-e5b73a86ca83.png)

## Additional context

- I had no issue creating node groups in the previous version of the module (v17)

**Comment by bryantbiggs** [comment]
Its very hard to view your code snippets without the proper formatting. However, this module does not set any explicit deny's so I would look elsewhere in  your environment - an SCP or permissions boundary perhaps?

**Comment by aozoeao** [comment]
I have only 2 policies attached to my user.
One that gives admin permissions to my user, dispensed by AWS.
`{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "*",
            "Resource": "*"
        }
    ]
}`
The other one to ensure i'm using MFA, which I am:
`    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iam:ChangePassword"
            ],
            "Resource": [
                "arn:aws:iam::*:user/${aws:username}"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "iam:GetAccountPasswordPolicy"
            ],
            "Resource": "*"
        },
        {
            "Sid": "AllowViewAccountInfo",
            "Effect": "Allow",
            "Action": "iam:ListVirtualMFADevices",
            "Resource": "*"
        },
        {
            "Sid": "AllowManageOwnVirtualMFADevice",
            "Effect": "Allow",
            "Action": [
                "iam:CreateVirtualMFADevice",
                "iam:DeleteVirtualMFADevice"
            ],
            "Resource": "arn:aws:iam::*:mfa/${aws:username}"
        },
        {
            "Sid": "AllowManageOwnUserMFA",
            "Effect": "Allow",
            "Action": [
                "iam:DeactivateMFADevice",
                "iam:EnableMFADevice",
                "iam:GetUser",
                "iam:ListMFADevices",
                "iam:ResyncMFADevice"
            ],
            "Resource": "arn:aws:iam::*:user/${aws:username}"
        },
        {
            "Sid": "DenyAllExceptListedIfNoMFA",
            "Effect": "Deny",
            "NotAction": [
                "iam:CreateVirtualMFADevice",
                "iam:EnableMFADevice",
                "iam:GetUser",
                "iam:ListMFADevices",
                "iam:ListVirtualMFADevices",
                "iam:ResyncMFADevice",
                "ecr:*",
                "sts:GetSessionToken",
                "iam:ChangePassword",
                "iam:GetAccountPasswordPolicy",
                "iam:DeleteVirtualMFADevice"
            ],
            "Resource": "*",
            "Condition": {
                "Bool": {
                    "aws:MultiFactorAuthPresent": "false"
                }
            }
        }
    ]
}`

**Comment by aozoeao** [comment]
I have tried creating node groups on another AWS account and it indeed worked.
However my questions reside in the fact that with the previous version of the eks module (v17) I had no issue creating node groups, what changed ? and what could be the root cause ?
Because to my understanding it's not a permission issue as I am admin.

**Comment by bryantbiggs** [comment]
> what changed?
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/UPGRADE-18.0.md

> So the error message displayed is not very helpful to me, perhaps i'm wrong?

We don't control the error messages here, thats in the provider or Terraform core

**Comment by aozoeao** [comment]
I have checked and there is no permissions boundary set for my user. SCP is disabled as well. Do you have any other leads?

I have checked changes mades in the v18 to `eks_managed_node_groups`. To my understanding the major difference to create node groups compared to the v17 is the absence of `for_each` and that they are now standalone sub-module used by the root `eks` module. Is that it?

**Comment by bryantbiggs** [comment]
closing for now since this issue is outside the scope of the module

**Comment by lnalex** [comment]
Posting in case anyone else comes across this issue:

This policy is likely the culprit:
```
        {
            "Sid": "DenyAllExceptListedIfNoMFA",
            "Effect": "Deny",
            "NotAction": [
                "iam:CreateVirtualMFADevice",
                "iam:EnableMFADevice",
                "iam:GetUser",
                "iam:ListMFADevices",
                "iam:ListVirtualMFADevices",
                "iam:ResyncMFADevice",
                "ecr:*",
                "sts:GetSessionToken",
                "iam:ChangePassword",
                "iam:GetAccountPasswordPolicy",
                "iam:DeleteVirtualMFADevice"
            ],
            "Resource": "*",
            "Condition": {
                "Bool": {
                    "aws:MultiFactorAuthPresent": "false"
                }
            }
        }
```

If you are using long term credentials (e.g. access key) this policy is not applicable (see https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa_configure-api-require.html#MFAProtectedAPI-overview). However, EKS needs to make an `iam:GetRole` call on your behalf as part of the `eks:CreateNodegroup` call.

This `GetRole` call made on your behalf is being made with temporary credentials, meaning that it will be denied unless you also made the `CreateNodegroup` call with temporary credentials that are also MFA authenticated (e.g. via GetSessionToken)

tl;dr fix by either removing this policy or use MFA authenticated temporary credentials to make the `CreateNodegroup` call/run terraform apply with.
